### ISSUE TITLE
Make reactions extendable

### ISSUE THREAD

**Issue by ischoegl** [Cantera/enhancements#63]
**Abstract**

At the moment, reactions are hard-coded and do not make use of 'factory' creation, i.e. they cannot be easily extended by user code compiled against core Cantera. Adding a `ReactionFactory` would introduce more flexibility. Further, reactions still use *magic numbers* that have been largely removed elsewhere.

**Motivation**

The main motivation would be to replace hard-coded reaction instantiation by an extendable `ReactionFactory` approach. 

Example: `GasKinetics` would become fully flexible, i.e. someone can define a new reaction type externally and link it against cantera (`ctapp` approach, where new C++ lambda functions are registered for an existing Cantera factory object).

**Possible Solutions**

Replace magic number referenced function calls by name-mapped C++ lambda functions calls (which are mainly used for setup; actual kinetics calculations involve templated functions, i.e. there would be no change from the current implementation whatsoever). Use C++ lambda functions to recycle existing code directly.

**References**

* Cantera/cantera#982 
* Cantera/cantera#745 (now closed), and corresponding [`ischoegl:add-reaction-factories`](https://github.com/ischoegl/cantera/tree/reaction-factories) branch
* [ctapp](https://github.com/ischoegl/ctapp) - illustration on how to create external objects and register them in existing Cantera class factories.
* Cantera/cantera#597 (no longer pursued) .. an example for a reaction type that is currently unavailable, and could be added in user code instead of core Cantera.

**Comment by ischoegl** [comment]
@speth ... a somewhat heretic (?) alternative would be to rewrite reaction models in Julia and including them via Julia’s C API. I honestly haven’t spent too much thought on it but it would have the advantage of using a compiled high level language that neatly integrates with data science. I am again tagging @jiweiqi here as I’m interested in the opinion of someone who has worked with on that area.

**Comment by speth** [comment]
The ability to write reaction models in Julia or Python is of course one of the goals of #79. My guess is that it would not be beneficial to only provide the existing models in Julia, though. Even though Julia is pretty fast, I think the overhead of moving between the languages would make it noticeably slower than sticking to C++ for functions that are called as many times as reaction rate evaluations. Also, I don't think I'd want to make Julia a required dependency of Cantera.

**Comment by ischoegl** [comment]
@speth ... I know it was heretic :wink:. You’re right about overhead, but I’d be interested in benchmarks. Perhaps @jiweiqi has ideas for wrapping Julia into C?

**Comment by jiweiqi** [comment]
To be honest, I don't have many ideas on how Cantera can move on this topic. I can say from my experience of developing two differentiable combustion packages: Arrhenius.jl in Julia and ReacTorch in PyTorch. One can ultimately view the interpreting of reaction mechanisms as data flow. All of the information is stored in the .YAML files. In my experience, writing an interpreter based on a human-readable structured YAML file is not so challenging. In ReacTorch, we re-wrote all of the interpreting code. In Arrhenius. jl, I use Cantera and ReacTorch to save the mechanism into a NumPy .npy file and load it to Julia.

I think all of those conveniences are coming from the efforts Cantera has made in formulating standard mechanism format. Therefore, to make reactions extendable, we can again look into YAML. As long as we have a good data source, the rest could be simple.

In practice, it depends on how much additional information we need in addition to what we already have in the YAML file. If we don't need that much, in Julia, we can manually code it and it is not so difficult. To me, the philosophy of Julia is more like Matlab, we don't wrap everything into a class, instead, everything is linked with functions. So that it is pretty straightforward to add customized reactions. That's also how we get neural network models in.

I personally don't use wrapping often since I have to use several computers. Configuring the environment is painful. I don't write C++ normally, but I don't think we should put Julia as a dependent for Cantera as well. Instead, if we feel C++ has made life not easy, we could re-write everything in Julia which has the same speed as C++. For zero-D reactors, I guess we could get everything we need within a couple of months. For one-D flame, I don't have much experience.

**Comment by ischoegl** [comment]
@jiweiqi ... thanks, it really looks like you haven't looked far beyond the YAML format. Overall, I'm not sure that we are really comparing apples to apples here. ReacTorch was started 10 months ago and hasn't seen a commit in 5 months. The first commit of Arrhenius.jl dates back 14 days as of today. As I have written elsewhere, getting a proof of concept for a small subset of capabilities off the ground is relatively simple and be can done pretty quickly (as you prove). I'm not sure that many of us are ready to discard things that are proven that easily. A lot of time goes into testing as cantera is production code for most of us.

**Comment by jiweiqi** [comment]
Totally agree. A side note about ReacTorch is that we don't push applications into the repo

**Comment by jiweiqi** [comment]
I am sorry for my inaccurate expressions. I definitely know the complexity of developing Cantera behind the scene. I tried to limit the discussions to "what we can do to make the reactions extendable and flexible for other high-level languages". Apologize for any confusion made in previous posts.

**Comment by ischoegl** [comment]
@speth ... Cantera/cantera#982 should facilitate testing of custom reactions that fall outside of a standardized YAML parameterization (in anticipation of #79). It is not mean to be a full rewrite; rather, it just repackages already available hard-coded reaction types in an extendable factory format (both XML and YAML nodes).

**Comment by speth** [comment]
@ischoegl, thank you for drafting a version of the `ReactionFactory` class. Seeing this helps clarify for me an issue that I tried to describe before, which is this: If you have written a new `Reaction` class that has a rate calculator class paralleling those in `RxnRates.h`, how do you add it to a `GasKinetics` object? Currently, the `GasKinetics` class has distinct member data structures for each type of reaction it handles, e.g. `Rate1<Plog> m_plog_rates`. So if you have a custom reaction rate class, then it needs to go in a different data structure, which means you need a new `Kinetics` class to support your new reaction. That would be sort of tolerable for just adding one new reaction type, but it would make it virtually impossible to support mixing and matching a range of user-defined reaction types.

I suspect, that the original performance rationale for doing reactions in this templated form, rather than using virtual functions, is outdated and doesn't apply to modern processors and C++ compilers, if it's done right. Instead, I think we can use a structure similar to what I implemented in the `MultiSpeciesThermo` a while back to eliminate a similarly rigid system that was being used there. Here, the reaction rate classes would all derive from an abstract base class, and implement a common set of functions. At its core, these functions would be:

1. A function to calculate and store any temperature-dependent parameters that would be the same for all reactions of the same type. For example, for Arrhenius-type reactions, this would just be calculating log(T).
2. A function to calculate and store any concentration/composition dependent parameters that would be the same for all reactions of the same type.
3. A function to calculate and store concentration/composition dependent terms specific to each reaction, using the stored data.
4. A function to calculate the reaction rate, using the stored data.

Then, in the `Kinetics` class, we would take the reactions, grouped by type, and call the first two functions as needed (i.e. in `update_rates_T` and `update_rates_C`, once for each unique reaction type only, and the latter two functions for each reaction. This structure would allow users to add new reactions that required any data they liked from the `ThermoPhase` or `Kinetics` object, which would only be queried once for the entire class of reactions. Furthermore, by grouping the reactions by type, I think the overhead of the virtual function calls will be minimized because calling all of the same specializations in a row will be easy for the branch predictor to optimize. When I did this for the species thermo classes, I tested the performance and wasn't able to identify any noticeable difference, so I think it's reasonable to expect the same here.

In addition to the increased flexibility, this approach would eliminate the need to have the `reg_addRxn` and `reg_modRxn` hooks that are currently introduced in Cantera/cantera#982.

**Comment by ischoegl** [comment]
@speth ... thanks for the detailed response. I agree that bypassing `RxnRates.h` and not adding to `GasKinetics` would complicate things, despite the obvious advantages. In my 'blueprint', I honestly just tried to describe the header file. I agree that a more flexible implementation will require some more extensive refactoring.

Looking at things overall, I believe the overall task can be split into two parts:

1. Creating a `ReactionFactory` that allows to add new reaction types (whether they are `GasKinetics`, `InterfaceKinetics` or even a more flexible successor)
2. Refactoring the code along the lines you describe above: it is absolutely true that my poor-man's version of PythonRate does not allow for concentration-dependent parameters to be configured on the C++ side, i.e. what #79 would require.

From my perspective, I don't think that the two steps need to happen at the same time. Step (1) already has a PR with Cantera/cantera#982, which should make step (2) more manageable as multiple approaches can be tested in parallel.

**Comment by ischoegl** [comment]
@speth ... as mentioned in Cantera/cantera#982, a draft introducing `ReactionFactory` along with a Python `CustomReaction` is ready.

**Comment by ischoegl** [comment]
Linking a [User Group discussion](https://groups.google.com/g/cantera-users/c/E8eSzSDclpE) here, with reactions that do not follow mass action kinetics.

**Comment by ischoegl** [comment]
At this point of Cantera/cantera#982, some benchmark results are added for an alternative implementation of reaction rates. The main difference to the current framework is that newly introduced `RxnRate` objects are no longer wrapped into templates as before, and thus become directly accessible.

```
Average time of 100 simulation runs for 'gri30.yaml' (CH4)
- Original mechanism: 148.14 ms (T_final=2736.60)
- One Python reaction: 156.36 ms (T_final=2736.60) ... +5.55%
- Alternative reactions: 150.46 ms (T_final=2736.60) ... +1.56%
```
It should be noted that Cantera/cantera#982 is a draft for an alternative framework, where changes are not fully implemented.
